onnxruntime/core/mlas/lib/sqnbitgemm_kernel_neon.cpp

```diff
@@ -0,0 +1,502 @@
+            UnrolledLoop<NCols>([&](size_t i) { b_data[i] += 8; });
+        }
+
+        // increment pointers to next block
+        QuantBData += MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
+        QuantBScale += 1;
+        QuantBZeroPointIdx += 1;
+    }
+
+    if constexpr (NCols == 4) {
+        float32x4_t sum = FoldAccumulators(acc[0], acc[1], acc[2], acc[3]);
+
+        if (BiasPtr != nullptr) {
+            sum = vaddq_f32(sum, vld1q_f32(BiasPtr));
+        }
+
+        vst1q_f32(SumPtr, sum);
+    } else {
+        for (size_t i = 0; i < NCols; ++i) {
+            SumPtr[i] = vaddvq_f32(acc[i]);
+            if (BiasPtr != nullptr) {
+                SumPtr[i] += BiasPtr[i];
+            }
+        }
+    }
+}
+
+}  // namespace
+
+//
+// MlasSQNBitGemmKernel and helpers.
+//
+
+template <size_t BlkBitWidth, size_t BlkLen>
+MLAS_FORCEINLINE size_t
+MlasSQNBitGemmKernelNeon(
+    const float* A,
+    const uint8_t* QuantBData,
+    const float* QuantBScale,
+    const uint8_t* QuantBZeroPoint,
+    float* C,
+    size_t CountM,
+    size_t CountN,
+    size_t CountK,
+    size_t lda,
+    size_t BlockStrideQuantB,
+    size_t ldc,
+    const float* Bias
+)
+{
+    [[maybe_unused]] auto impl0_reference = [&]() {
+        static_assert(BlkBitWidth == 4);
+
+        for (size_t m = 0; m < CountM; ++m) {
+            for (size_t n = 0; n < CountN; ++n) {
+                for (size_t k = 0, k_blk_idx = 0; k < CountK; k += BlkLen, k_blk_idx += 1) {
+                    const size_t kblocklen = std::min(CountK - k, BlkLen);
+
+                    const float b_s = QuantBScale[n * BlockStrideQuantB + k_blk_idx];
+                    const uint8_t b_z = [&]() -> uint8_t {
+                        if (QuantBZeroPoint != nullptr) {
+                            const size_t i = n * BlockStrideQuantB + k_blk_idx;
+                            uint8_t zp_packed = QuantBZeroPoint[i / 2];
+                            return ((i & 1) == 1) ? (zp_packed >> 4) : (zp_packed & 0x0F);
+                        } else {
+                            return 8;
+                        }
+                    }();
+                    const uint8_t* b_data =
+                        QuantBData + (n * BlockStrideQuantB + k_blk_idx) * MlasQNBitBlkDataSizeInBytes(BlkBitWidth, BlkLen);
+
+                    for (size_t kk = 0; kk < kblocklen; ++kk) {
+                        uint8_t b_packed = b_data[kk / 2];
+                        uint8_t b_byte = ((kk & 1) == 1) ? (b_packed >> 4) : (b_packed & 0x0F);
+                        float b_value = (b_byte - b_z) * b_s;
+
+                        C[m * ldc + n] += A[m * lda + k + kk] * b_value;
+                    }
+                }
+
+                if (Bias) {
+                    C[m * ldc + n] += Bias[n];
+                }
+            }
+        }
+
+        return CountM;
+    };
+
+    auto impl1 = [&]() {
+        constexpr size_t NCols = 4;
```

Comment:
>NCols [](http://example.com/codeflow?start=25&length=5)

Ncols is too small here. For llama 7b model, K can be as big as 11008 thus 1 row of A takes ~44,032 bytes. You need to check the L1 cache size of the target ARM CPU. In a big chance, A will be flushed for each NCols. That is for each load of A to cache, there is only 4 computation. it is too low. You can either:
1. split K to CountK and load a tile [CountN, CountK] to cache and then it doesn't matter much to use a small NCols here. You can refer to [https://github.com/microsoft/onnxruntime/blob/c250540722a8028250de1d572365ec73ef6ecedf/onnxruntime/core/mlas/lib/sgemm.cpp#L1140](https://github.com/microsoft/onnxruntime/blob/c250540722a8028250de1d572365ec73ef6ecedf/onnxruntime/core/mlas/lib/sgemm.cpp#L1140)
Or 2. you can try to increase NCols, which is limited by the register numbers.

Reply:
I tried 2. without much success.
for k/n = 4096/11008, 11008/4096, and 4096/4096, NCols = 8 is similar in performance, NCols = 16 got significantly worse. the code size also increases with larger NCols values.

I can experiment with 1. but this is a larger change. perhaps in a future PR?

Reply:
This looks like optimized for M==1, where you simply scan multiple columns of B matrix once.  For cases where M is greater, when processing first row of A, you scan multiple columns of B once. For the next row, you scan the same columns of B again.

One thing you can try is partition on the K dimension, as Yufeng mentioned. Also, maybe try loading multiple row of A instead of just one. For example,  define MRow, and process MRow of A and NCol of B in one iteration.